src/orange/orangeinp/detail/DeMorganSimplifier.cc

```diff
@@ -0,0 +1,443 @@
+/*!
+ * Construct and fix bitset size.
+ */
+DeMorganSimplifier::DeMorganSimplifier(CsgTree const& tree) : tree_(tree)
+{
+    orphaned_nodes_.resize(tree_.size());
+    new_negated_nodes_.resize(tree_.size());
+    simplified_negated_nodes_.resize(tree_.size());
+    negated_join_nodes_.resize(tree_.size());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Perform the simplification. The state of the instance isn't cleared, so only
+ * call this once.
+ */
+CsgTree DeMorganSimplifier::operator()()
+{
+    find_join_negations();
+    return build_simplified_tree();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * During the first pass, a negated join operation is by default flag for
+ * removal during the construction of the simplified tree. If we later discover
+ * another parent that needs to keep the \c Joined node, this function removes
+ * the \c Joined node and its descendents from the list of orphans.
+ *
+ * \param node_id the node to keep and its descendents.
+ */
+void DeMorganSimplifier::record_parent_for(NodeId node_id)
+{
+    std::visit(
+        Overload{
+            [&](Joined const& joined) {
+                // each operand of the Joined node also needs to be
+                // kept
+                for (auto join_operand : joined.nodes)
+                {
+                    record_parent_for(join_operand);
+                }
+                // and the Joined node itself needs to be kept
+                CELER_EXPECT(std::holds_alternative<Joined>(tree_[node_id]));
+                orphaned_nodes_[node_id.get()] = false;
+            },
+            [&](Negated const& negated) {
+                // A Negated node with a Joined child will get removed
+                // regardless of being flagged as an orphan so don't unflag its
+                // child.
+                CELER_EXPECT(std::holds_alternative<Negated>(tree_[node_id]));
+                if (!std::get_if<Joined>(&tree_[negated.node]))
+                {
+                    record_parent_for(negated.node);
+                    orphaned_nodes_[node_id.get()] = false;
+                }
+            },
+            // other nodes don't have children and can never be flagged as
+            // orphan
+            [](auto&&) {}},
+        tree_[node_id]);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Declare a \c Negated node with a \c Joined child that must be transformed to
+ * the opposite join operation. Per DeMorgan's law, this will also make sure
+ * that all operands of the join operation have a \c Negated parent added.
+ *
+ * \param negated_id of a \c Negated node pointing to \c Joined node inside
+ * the tree_.
+ */
+void DeMorganSimplifier::record_join_negation(NodeId negated_id)
+{
+    CELER_EXPECT(negated_id < tree_.size());
+
+    CELER_ASSERT(std::holds_alternative<Negated>(tree_[negated_id]));
```

Comment:
```suggestion
    CELER_EXPECT(std::holds_alternative<Negated>(tree_[negated_id]));
```
This one is a precondition on input :) You can also use `CELER_ASSUME` to help the compiler optimize when assertions are disabled.